src/main/java/io/confluent/copycat/hdfs/HdfsWriter.java

```diff
@@ -72,95 +95,152 @@ public HdfsWriter(HdfsSinkConnectorConfig connectorConfig) {
+          public void run() {
+            lock.lock();
+            try {
+              rotate(topicPart);
+            } catch (IOException | InterruptedException e) {
+              throw new CopycatRuntimeException(e);
+            } finally {
+              lock.unlock();
+            }
+          }
+        }, rotateInterval, TimeUnit.SECONDS);
+        tasks.put(topicPart, task);
+      }
       if (shouldRotate(topicPart)) {
         rotate(topicPart);
       }
     } catch (IOException e) {
       throw new CopycatRuntimeException(e);
+    } finally {
+      lock.unlock();
     }
   }
 
-  public void close() throws IOException, CopycatException {
-    for (TopicPartition topicPart: writers.keySet()) {
+  public void close() throws IOException, CopycatException, InterruptedException {
+    timedRotateThread.shutdown();
+
+    for (TopicPartition topicPart : writers.keySet()) {
       rotate(topicPart);
     }
     writers.clear();
+
+    for (WAL wal : wals.values()) {
+      wal.close();
+    }
+    wals.clear();
+
+    tempFileNames.clear();
+    recordCounters.clear();
+    offsets.clear();
+    tasks.clear();
     if (fs != null) {
       fs.close();
     }
   }
 
-  private Map<TopicPartition, Long> readOffsets() throws IOException {
-    Map<TopicPartition, Long> offsets = new HashMap<TopicPartition, Long>();
+  private void recover(TopicPartition topicPart) throws IOException {
+    if (!recovered.contains(topicPart)) {
+      WAL wal = new WAL(conf, url, topicsDir, topicPart);
+      wal.apply();
+      wals.put(topicPart, wal);
+      recovered.add(topicPart);
+      readOffsets(topicPart);
+    }
+  }
+
+  private void readOffsets() throws IOException {
     Path path = new Path(url + "/" + topicsDir);
-    PathFilter filter = new CommittedFileFilter();
     FileStatus[] topics = fs.listStatus(path);
-    for (FileStatus topic: topics) {
+    for (FileStatus topic : topics) {
       if (topic.isDirectory()) {
         String topicName = topic.getPath().getName();
-        FileStatus[] partitions = fs.listStatus(topic.getPath(), filter);
+        FileStatus[] partitions = fs.listStatus(topic.getPath());
         for (FileStatus partition : partitions) {
-          if (partition.isFile()) {
-            String filename = partition.getPath().getName();
-            String[] parts = filename.split("\\.");
-            try {
-              int partitionNum = Integer.parseInt(parts[0]);
-              long offset = Long.parseLong(parts[1]);
-              TopicPartition topicPart = new TopicPartition(topicName, partitionNum);
-              if (!offsets.containsKey(topicPart) || offset > offsets.get(topicPart)) {
-                offsets.put(topicPart, offset);
-              }
-            } catch (NumberFormatException e) {
-              log.warn("Invalid Copycat HDFS file: {}", filename);
-            }
+          if (partition.isDirectory()) {
+            int partitionNum = Integer.parseInt(partition.getPath().getName());
```

Comment:
What about a `NumberFormatException` if somebody puts some other directory in here with the normal Copycat data?